src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1962,6 +1963,151 @@ static TypeDeclarationSyntax constructTypeDeclaration(ContextAwareSyntax syntaxF
             }
         }
 
+        private MemberDeclarationSyntax ParseExtensionContainer(SyntaxList<AttributeListSyntax> attributes, SyntaxListBuilder modifiers)
+        {
+            Debug.Assert(!IsInAsync);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.ExtensionKeyword);
+            SyntaxToken extensionKeyword = this.EatContextualToken(SyntaxKind.ExtensionKeyword);
+
+            var outerSaveTerm = _termState;
+            _termState |= TerminatorState.IsPossibleAggregateClauseStartOrStop; // TODO2 add test affected
+
+            TypeParameterListSyntax typeParameters = this.ParseTypeParameterList();
+            ReceiverParameterSyntax receiverParameter = parseReceiverParameter(out var openParenToken, out var closeParenToken);
```

Comment:
parse a parameter list, with a requirement that at least a single parameter is required.  This can be done by updating ParseParameterList to take a parameter stating if a single parameter is required (and passing that to hte hlper it calls).

you can optionally error at parse time or binding time if there is >1 parameter.  I recommend binding time.

Reply:
By requiring a single parameter, you can be sure in later phases that it is safe to access [0]